float32 vec_x
float32 vec_y
float32 vec_z